targets/TARGET_Cypress/TARGET_PSOC6/TARGET_WHD/src/bus_protocols/whd_bus_sdio_protocol.c

```diff
@@ -0,0 +1,1427 @@
+
+    if (mode == SDIO_BYTE_MODE)
+    {
+        whd_assert("whd_bus_sdio_cmd53: data_size > 512 for byte mode", (data_size <= (uint16_t )512) );
+        arg.cmd53.count = (unsigned int)(data_size & 0x1FF);
+
+        result =
+            cyhal_sdio_bulk_transfer(whd_driver->bus_priv->sdio_obj, (cyhal_transfer_t)direction, arg.value,
+                                     (uint32_t *)data, data_size, response);
+
+        if (result != CY_RSLT_SUCCESS)
+        {
+            WPRINT_WHD_ERROR( ("%s:%d cyhal_sdio_bulk_transfer SDIO_BYTE_MODE failed\n", __func__, __LINE__) );
+            goto done;
+        }
+    }
+    else
+    {
+        arg.cmd53.count = (unsigned int)( (data_size / (uint16_t)SDIO_64B_BLOCK) & BIT_MASK(9) );
+        if ( (uint32_t)(arg.cmd53.count * (uint16_t)SDIO_64B_BLOCK) < data_size )
+        {
+            ++arg.cmd53.count;
+        }
+        arg.cmd53.block_mode = (unsigned int)1;
+
+        result =
+            cyhal_sdio_bulk_transfer(whd_driver->bus_priv->sdio_obj, (cyhal_transfer_t)direction, arg.value,
+                                     (uint32_t *)data, data_size, response);
+
+        if (result != CY_RSLT_SUCCESS)
+        {
+            WPRINT_WHD_ERROR( ("%s:%d cyhal_sdio_bulk_transfer failed\n", __func__, __LINE__) );
+            goto done;
+        }
+    }
+
+    if (direction == BUS_READ)
+    {
+        WHD_BUS_STATS_INCREMENT_VARIABLE(whd_driver->bus_priv, cmd53_read);
+        add_log_entry(direction, function, address, data_size, data);
+    }
+
+done:
+    WHD_BUS_STATS_CONDITIONAL_INCREMENT_VARIABLE(whd_driver->bus_priv,
+                                                 ( (result != WHD_SUCCESS) && (direction == BUS_READ) ),
+                                                 cmd53_read_fail);
+    WHD_BUS_STATS_CONDITIONAL_INCREMENT_VARIABLE(whd_driver->bus_priv,
+                                                 ( (result != WHD_SUCCESS) && (direction == BUS_WRITE) ),
+                                                 cmd53_write_fail);
+    if (result == CY_RSLT_SUCCESS)
+        return WHD_SUCCESS;
+    else
+        return WHD_HAL_ERROR;
+}
+
+static whd_result_t whd_bus_sdio_download_firmware(whd_driver_t whd_driver)
+{
+    uint8_t csr_val = 0;
+    whd_result_t result;
+    uint32_t loop_count;
+    uint32_t ram_start_address;
+
+    ram_start_address = GET_C_VAR(whd_driver, ATCM_RAM_BASE_ADDRESS);
+
+    if (ram_start_address != 0)
+    {
+        CHECK_RETURN(whd_reset_core(whd_driver, WLAN_ARM_CORE, SICF_CPUHALT, SICF_CPUHALT) );
+    }
+    else
+    {
+        CHECK_RETURN(whd_disable_device_core(whd_driver, WLAN_ARM_CORE, WLAN_CORE_FLAG_NONE) );
+        CHECK_RETURN(whd_disable_device_core(whd_driver, SOCRAM_CORE, WLAN_CORE_FLAG_NONE) );
+        CHECK_RETURN(whd_reset_device_core(whd_driver, SOCRAM_CORE, WLAN_CORE_FLAG_NONE) );
+
+        CHECK_RETURN(whd_chip_specific_socsram_init(whd_driver) );
+    }
+
+#if 0
```

Comment:
Is this code needed anymore? Could it be put behind a flag?

Reply:
This code is used internally to debug SDIO bus transfers. It is not related to Mbed.